src/store_client.cc

```diff
@@ -499,58 +517,126 @@ store_client::fileRead()
     assert(!flags.disk_io_pending);
     flags.disk_io_pending = true;
 
+    // mem->swap_hdr_sz is zero here during initial read(s)
+    const auto nextStoreReadOffset = NaturalSum<int64_t>(mem->swap_hdr_sz, nextHttpReadOffset()).value();
+
+    // XXX: If fileRead() is called when we do not yet know mem->swap_hdr_sz,
+    // then we must read from disk offset zero to learn it. We cannot compute
+    // correct HTTP response start offset on disk without it. However, late
+    // startSwapin() calls imply that the assertion below may fail.
+    Assure(mem->swap_hdr_sz > 0 || !nextStoreReadOffset);
```

Comment:
Please evaluate this XXX: Do you agree that there is a problem here? Do you agree that essentially the same problem exists in the official code as well? Can you think of a way to trigger this assertion in a test? Among all the requests in this review, these questions have the highest priority.

Reply:
I could not find a theoretical scenario (or a test) how to trigger this problem here. The condition is somewhat equivalent to the situation when we were reading from memory (while detached from disk) and then decided to start reading from disk.

store_client::fileRead() may be called in the following cases:
1. Response is being received from the server.
  a. In the Store writer itself swap_hdr_sz is assigned early (by storeSwapOutStart() and getSerialisedMetaData()) and before store_client::fileRead() when the entry is firstly written to the Store). 
  b. In a different worker, the disk entry is not available until it has been written on disk.
2. Response is being served from the Store.  
  a. Disk entry, swap_hdr_sz is initialized synchronously by store_client::readHeader() and Store::UnpackHitSwapMeta() just after store_client::fileRead() is first called.
  b. Shared memory entry. swap_hdr_sz remains 0, but we must not get into store_client::fileRead().

So, if we get into store_client::fileRead() after a series of readings from local memory it should be OK because this local entry must be attached to the disk entry.
The condition could have been possible if we could start reading from the shared memory and the switch to disk. I think this is impossible in the current code.